updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,218 @@
+    nodes: Optional[Dict[str, Optional[str]]] = None
+    max_assigned_nodes: int = 0
+
+
+@dataclass(kw_only=True)
+class ClusterManagerInput:
+    state: Optional[ClusterManagerState] = None
+    test_continue_as_new: bool = False
+
+
+@dataclass
+class ClusterManagerResult:
+    max_assigned_nodes: int
+    num_currently_assigned_nodes: int
+
+
+@dataclass(kw_only=True)
+class ClusterManagerAllocateNNodesToJobInput:
+    num_nodes: int
+    task_name: str
+
+
+@dataclass(kw_only=True)
+class ClusterManagerDeleteJobInput:
+    task_name: str
+
+
+# ClusterManagerWorkflow keeps track of the allocations of a cluster of nodes.
+# Via signals, the cluster can be started and shutdown.
+# Via updates, clients can also assign jobs to nodes and delete jobs.
+# These updates must run atomically.
+@workflow.defn
+class ClusterManagerWorkflow:
+    def __init__(self) -> None:
+        self.state = ClusterManagerState()
+        # Protects workflow state from interleaved access
+        self.nodes_lock = asyncio.Lock()
+
+    @workflow.signal
+    async def start_cluster(self):
+        self.state.cluster_started = True
+        self.state.nodes = {str(k): None for k in range(25)}
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        self.state.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(
+        self, input: ClusterManagerAllocateNNodesToJobInput
+    ) -> List[str]:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        if self.state.cluster_shutdown:
+            # If you want the client to receive a failure, either add an update validator and throw the
+            # exception from there, or raise an ApplicationError. Other exceptions in the main handler
+            # will cause the workflow to keep retrying and get it stuck.
+            raise ApplicationError(
+                "Cannot allocate nodes to a job: Cluster is already shut down"
+            )
+
+        async with self.nodes_lock:
+            unassigned_nodes = [k for k, v in self.state.nodes.items() if v is None]
+            if len(unassigned_nodes) < input.num_nodes:
+                # If you want the client to receive a failure, either add an update validator and throw the
+                # exception from there, or raise an ApplicationError. Other exceptions in the main handler
+                # will cause the workflow to keep retrying and get it stuck.
+                raise ApplicationError(
+                    f"Cannot allocate {input.num_nodes} nodes; have only {len(unassigned_nodes)} available"
+                )
+            assigned_nodes = unassigned_nodes[: input.num_nodes]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
```

Comment:
Would help users understand the locking even more if this comment said what it is that shouldn't be interleaved.